packages/buckets/src/api/index.ts

```diff
@@ -542,42 +578,63 @@ export async function bucketsPushPath(
             },
             root: event.root?.path ?? '',
           }
-          resolve(res)
+          return resolve(res)
         } else if (opts?.progress) {
           opts.progress(event?.bytes)
         }
       } else {
-        reject(new Error('Invalid reply'))
+        return reject(new Error('Invalid reply'))
       }
     })
     client.onEnd((code, msg) => {
       if (code !== grpc.Code.OK) {
         const message = msg ? msg : code.toString()
-        reject(new Error(message))
+        return reject(new Error(message))
       } else {
-        resolve()
+        return resolve()
       }
     })
     if (source) {
       const head = new PushPathRequest.Header()
       head.setPath(source.path || path)
       head.setKey(key)
+      // Setting root here ensures pushes will error if root is out of date
+      let root = ''
+      if (opts?.root) {
+        // If we explicitly received a root argument, use that
+        root = typeof opts.root === 'string' ? opts.root : opts.root.path
+      } else {
+        // Otherwise, make a call to list path to get the latest known root
+        const head = await bucketsListPath(api, key, '', ctx)
+        root = head.root?.path ?? '' // Shouldn't ever be undefined here
+      }
+      head.setRoot(root)
       const req = new PushPathRequest()
       req.setHeader(head)
       const metadata = { ...api.context.toJSON(), ...ctx?.toJSON() }
+      // Let's just make sure we haven't aborted this outside this function
```

Comment:
Just in case, we might as well just reject on not even start the thing if we've already aborted for some reason.

Reply:
This is useful in cases where we might have a bunch of pushes all queued up.